azurelinuxagent/common/utils/archive.py

```diff
@@ -0,0 +1,213 @@
+agent state is flushed whenever there is a incarnation change.
+The flush is archived periodically (once an hour).
+
+The process works as follows whenever a new incarnation arrives.
+
+ 1. Flush - move all state files to a new directory under
+ .../history/timestamp/.
+ 2. Archive - enumerate all directories under .../history/timestamp
+ and create a .zip file named timestamp.zip.  Delete the archive
+ directory
+ 3. Purge - glob the list .zip files, sort by timestamp in descending
+ order, keep the first 50 results, and delete the rest.
+
+... is the directory where the agent's state resides, by default this
+is /var/lib/waagent.
+
+The timestamp is an ISO8601 formatted value.
+"""
+
+MAX_ARCHIVED_STATES = 50
+
+CACHE_PATTERNS = [
+    re.compile("^(.*)\.(\d+)\.(agentsManifest)$", re.IGNORECASE),
+    re.compile("^(.*)\.(\d+)\.(manifest\.xml)$", re.IGNORECASE),
+    re.compile("^(.*)\.(\d+)\.(xml)$", re.IGNORECASE)
+]
+
+# 2018-04-06T08:21:37.142697
+# 2018-04-06T08:21:37.142697.zip
+ARCHIVE_PATTERNS_DIRECTORY = re.compile('^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+$')
+ARCHIVE_PATTERNS_ZIP       = re.compile('^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+\.zip$')
+
+
+class StateFlusher(object):
+    def __init__(self, lib_dir):
+        self._timestamp = datetime.now()
+        self._source = lib_dir
+        self._target = os.path.join(self._source, 'history', self._timestamp.isoformat())
+
+    def flush(self):
+        files = self._get_files_to_archive()
+        if len(files) == 0:
+            return
+
+        if self._mkdir():
+            self._archive(files)
+        else:
+            self._purge(files)
+
+    def _get_files_to_archive(self):
+        files = []
+        for f in os.listdir(self._source):
+            full_path = os.path.join(self._source, f)
+            for pattern in CACHE_PATTERNS:
+                m = pattern.match(f)
+                if m is not None:
+                    files.append(full_path)
+                    break
+
+        return files
+
+    def _archive(self, files):
+        for file in files:
+            fileutil.move_file(file, to_dir=self._target)
+
+    def _purge(self, files):
+        for file in files:
+            os.remove(file)
+
+    def _mkdir(self):
+        try:
+            fileutil.mkdir(self._target, mode=0o700)
+            return True
+        except:
+            return False
+
+
+# @total_ordering first appeared in Python 2.7 and 3.2
+# If there are more use cases for @total_ordering, I will
+# consider re-implementing it.
+class State(object):
+    def __init__(self, path, timestamp):
+        self._path = path
+        self._timestamp = timestamp
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    def delete(self):
+        pass
+
+    def archive(self):
+        pass
+
+    def __eq__(self, other):
+        return self._timestamp == other.timestamp
+
+    def __ne__(self, other):
+        return self._timestamp != other.timestamp
```

Comment:
Nit: Dual relops are often defined by stating one, then stating the other as the negation of the first. That is...
```
def __ne__(self, other):
    return not self == other
```
Similarly, __le__ is defined as not > and __ge__ is defined as not <.

Reply:
Agreed, but I also think it reads clearer when all of the operators are inline.  Your eyes can just scroll down and see how ever comparison is made.

Reply:
Which means the eyes glaze over and a flip of < for > or a missing (or extra) = is easy to miss. This may very well be a language-specific style; the C++ library style is very much as I described. I'm definitely not hardcore about this; having pushed back once, I'm done. :-)